### DISCUSSION TITLE
Element Exporter

### DISCUSSION THREAD

**Discussion by andreas83** [craftcms/cms#12476] [Ideas] [answered]
Recently i started to play with the ElementExporter, while doing so i noticed a few design flaws:

1. CraftCMS uses associative arrays, but for unknown reasons the sorting of the array position in honored,  not the array key itself. 
```
$inputArray = [];
$inputArray[0]["id"] = "id1";
$inputArray[0]["title"] = "title1";
$inputArray[0]["test"] = "test1";


$inputArray[1]["id"] = "id2";
$inputArray[1]["test"] = "test2";
$inputArray[1]["title"] = "title2";


//Result:  test2 is inside title column of csv. 
//Expected Result: every value is in their corresponding column 
```

2. I would love to add another Formater (ie. excel), sadly its only documented how to writer a Custom ElementExporter but not formater. 

Bonus: 
The current csv formatter produces sometimes invalid csv. I had some html inside one cell and it breaked the csv, which is most likely a bug. i will create a seperate bug report for this later.

**Comment by brandonkelly** [comment, marked as answer]
The key sorting thing is just how PHP arrays work. If you want an array to be resorted by its key, you will need to pass it to [`ksort()`](https://www.php.net/manual/en/function.ksort.php):

```php
ksort($inputArray);
```

It’s not possible to add a custom formatted at this time. We will be completely rewriting how exporting works down the road, which will make it possible to have full control over the export format.

As for the bug, please send your Composer files and a DB dump to support@craftcms.com, and point out what you exported, so we can try to reproduce.

**Reply by mihob** [reply]
Hello @brandonkelly ,

has anything happened here in the meantime? It would be really great if you had the option of implementing your own export formats.  Our editors usually can't do much with a CSV file, Excel would be much more suitable here

**Reply by brandonkelly** [reply]
Excel supports CSV

**Reply by mihob** [reply]
Yes... it supports CSV, but the default separator is tabstop and if you don't know how to change it correctly, Excel will not open the exports correctly.

And it is very tedious to constantly guide editors who are not so tech-savvy through the process.
